File: gev_reg_m.R
rm(list = ls())
library(evd)
# multiple GEV optim for the global trend
  # multiple sites
  # tvec( stationary: site1_para, site2_para, ..., 
#        nonstationary: common reg model)
  
  n = 100
  true_theta=c(100,30,0.1)
  p= 10
  true_beta = rep(0,p)
  # optim control
  optim_controlList = list()
  optim_controlList$maxit = 1e+3
  
  
  set.seed(1)
  xlist = zlist = list()
  ns = 10
  
  for (i in 1:ns)
  {
    z = matrix(rnorm(n*p), n, p)
    eps = rgev(n,loc=true_theta[1], 
               scale=true_theta[2], 
               shape=true_theta[3])
    xlist[[i]] = z%*%true_beta + eps 
    zlist[[i]] = z
  }
  

  gevreg_m = function(xlist, zlist, lambda = 0)
  {
    p = ncol(zlist[[1]])
    tvec = rep(0, ns*3 + p)
    ns = length(xlist)
    
    
    l2gev_m = function (tvec, lambda, xlist, zlist)
    {
      ns = length(xlist)
      v1 = 0
      for ( i in 1: length(xlist))
      {
        x = xlist[[i]]
        z = zlist[[i]]
        loc.vec.reg = drop(z%*%tail(tvec, p))
        loc.vec = tvec[3*(i-1)+1] + loc.vec.reg
        sc.vec = tvec[3*(i-1)+2]
        sh.vec = tvec[3*(i-1)+3]
        v1 = v1 - sum(lgev(x, loc = loc.vec, 
                           scale = sc.vec, shape = sh.vec))  
      }
      # loglikelihood
      
      # regularization
      v2 = lambda*sum(tail(tvec,p)^2)
      v = v1 + v2
      return(v)
    }
    
    lgev = function (x, loc = 0, scale = 1, shape = 0) 
    {
      if (min(scale) <= 0) 
        return( - 1e+6)
      if (length(shape) != 1) 
        stop("invalid shape")
      x <- (x - loc)/scale
      if (shape == 0) 
        d <- log(1/scale) - x - exp(-x)
      else {
        nn <- length(x)
        xx <- 1 + shape * x
        xxpos <- xx[xx > 0 | is.na(xx)]
        scale <- rep(scale, length.out = nn)[xx > 0 | is.na(xx)]
        d <- numeric(nn)
        d[xx > 0 | is.na(xx)] <- log(1/scale) - xxpos^(-1/shape) - 
          (1/shape + 1) * log(xxpos)
        d[xx <= 0 & !is.na(xx)] <- -(1e+6)
      }
      return(d)
    }
    
    
    for ( i in 1:ns)
    {
      x = xlist[[i]]
      start <- list()
      start$scale <- sqrt(6 * var(x))/pi
      start$loc <- mean(x) - 0.58 * start$scale
      tvec[3*(i-1)+1] = start$loc
      tvec[3*(i-1)+2] = start$scale
    }
    
    
    return( optim(tvec, l2gev_m, lambda = lambda, 
                  method = "BFGS", 
                  control = optim_controlList,
                  xlist = xlist,
                  zlist = zlist)$par) 
  }
  
 fitvec = gevreg_m(xlist, zlist, lambda = 1)
 matrix(head(fitvec, 3*ns), ns, 3, byrow = T)
 tail(fitvec,p)

File: gev_reg_m.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

import numpy as np
from scipy.optimize import minimize
from scipy.stats import genextreme

n = 3
ns = 2
p = 2
true_theta = np.array([100,30,0.1], dtype = float)
true_beta = np.repeat(0.0,p)
xlist = []
zlist = []
for i in range(ns):
    z = np.random.normal(size = n*p)
    z = z.reshape(n,p)
    x = genextreme.rvs(loc = true_theta[0],
               scale = true_theta[1],
               c = true_theta[2],
               size=n)
    xlist.append(x)
    zlist.append(z)

def gevreg_m(xlist, zlist, lambda =0 ):
    p = zlist[1].shape[1]
    ns = len(xlist)
    tvec = np.repeat(0.0, ns*3+p)
    

    def lgev(x, loc = 0, scale = 1, shape = 0):
        if (scale <= 0) :
            return ( -1e+6)
        x = (x - loc)/scale
        if (shape == 0):
            d = np.log(1/scale) -x -np.exp(-x)
        else:
            nn = len(x)
            xx = 1 + shape*x
            xxpos = xx[xx >0 | np.isnan(x)]
            scale = np.repeat(scale, nn)[ (xx>0) |(np.isnan(x))]
            d =  np.repeat(0.0, nn)
            d[xx > 0 | np.isnan(xx)] =  np.log(1/scale)  \
            - np.power(xxpos,(-1/shape)) \
            -(1/shape + 1) * np.log(xxpos)
            d[(xx <= 0) & (~np.isnan(xx))] = -(1e+6)
        return(d)
    

    x = xlist[1]        
    loc = 100.0
    scale = 20.0
    shape = 0.1    
    lgev(xlist[1], loc = loc, scale = scale, shape = shape)
    lam = 1.0
    # 



    for i in range(ns):
        x = xlist[i]
        start_scale = np.sqrt(6.0 * np.var(x))/np.pi
        start_loc = np.mean(x) - 0.58 * start_scale
        tvec[3*i] = start_loc
        tvec[3*i+1] = start_scale
    
    
    args = {'lam': lam, 'xlist': xlist, 'zlist':zlist}
    
    
    def l2gev_m (tvec, args):
        lam, xlist, zlist = args['lam'], args['xlist'], args['zlist']
        ns = len(xlist)
        v1 = 0
        for i in range(ns):
            x = xlist[i]
            z = zlist[i]
            loc_vec_reg = np.matmul(z,tvec[-p:(ns*3+p)])
            loc_vec = tvec[3*i] + loc_vec_reg
            sc_vec = tvec[3*i+1]
            sh_vec = tvec[3*i+2]
            v1 = v1 - np.sum(lgev(x, loc = loc_vec, scale = sc_vec,
                           shape = sh_vec))
        v2 =  lam*np.sum(np.power(tvec[-p:],2))
        v = v1 + v2
        return (v)
    
    l2gev_m (tvec, args)
  
    fit = minimize(l2gev_m, x0 = tvec, method='BFGS',args = args )

